File: frontend/src/types/cnis.ts
export type StatusProcessamento = "pendente" | "processando" | "concluido" | "erro";

export type TipoBeneficio =
  | "aposentadoria_idade_urbana"
  | "aposentadoria_idade_rural"
  | "aposentadoria_tempo_contribuicao"
  | "aposentadoria_especial_15"
  | "aposentadoria_especial_20"
  | "aposentadoria_especial_25"
  | "aposentadoria_pcd_idade"
  | "aposentadoria_pcd_tempo"
  | "auxilio_doenca"
  | "aposentadoria_invalidez"
  | "salario_maternidade"
  | "pensao_morte";

export const TIPOS_BENEFICIO: Record<TipoBeneficio, string> = {
  aposentadoria_idade_urbana: "Aposentadoria por Idade (Urbana)",
  aposentadoria_idade_rural: "Aposentadoria por Idade (Rural)",
  aposentadoria_tempo_contribuicao: "Aposentadoria por Tempo de Contribuição",
  aposentadoria_especial_15: "Aposentadoria Especial (15 anos)",
  aposentadoria_especial_20: "Aposentadoria Especial (20 anos)",
  aposentadoria_especial_25: "Aposentadoria Especial (25 anos)",
  aposentadoria_pcd_idade: "Aposentadoria PCD por Idade",
  aposentadoria_pcd_tempo: "Aposentadoria PCD por Tempo",
  auxilio_doenca: "Auxílio-Doença",
  aposentadoria_invalidez: "Aposentadoria por Invalidez",
  salario_maternidade: "Salário-Maternidade",
  pensao_morte: "Pensão por Morte",
};

export interface CNISListItem {
  id: string;
  cliente_id: string;
  nis: string;
  periodo_inicial_cn: string | null;
  periodo_final_cn: string | null;
  tempo_contribuicao_anos: number | null;
  total_contribuicoes: number | null;
  status_processamento: StatusProcessamento;
  created_at: string;
}

export interface CNISListResponse {
  items: CNISListItem[];
  total: number;
  skip: number;
  limit: number;
}

export interface PeriodoContribuicao {
  id: string;
  cnpj_empregador: string | null;
  razao_social_empregador: string | null;
  data_inicio: string;
  data_fim: string | null;
  tipo_contribuinte: string | null;
  dias_contribuidos: number;
  indicador_aposentadoria_especial: boolean;
  agente_nocivo: string | null;
  grau_deficiencia: string | null;
  total_remuneracoes: number | null;
  quantidade_remuneracoes: number | null;
  periodo_valido: boolean;
  motivo_invalidacao: string | null;
}

export interface CNIS {
  id: string;
  tenant_id: string;
  cliente_id: string;
  nis: string;
  data_nascimento: string;
  periodo_inicial_cn: string | null;
  periodo_final_cn: string | null;
  tempo_contribuicao_total_dias: number | null;
  tempo_contribuicao_anos: number | null;
  total_contribuicoes: number | null;
  maior_salario_contribuicao: number | null;
  media_salarios_contribuicao: number | null;
  status_processamento: StatusProcessamento;
  erros_validacao: Record<string, unknown> | null;
  created_at: string;
}

export interface InconsistenciaItem {
  tipo: string;
  descricao: string;
  periodo_afetado: string | null;
  impacto_financeiro: number | null;
  recomendacao: string | null;
}

export interface AnaliseInconsistencias {
  cnis_id: string;
  total_inconsistencias: number;
  inconsistencias: InconsistenciaItem[];
  periodos_sobrepostos: Array<{ tipo: string; descricao: string; periodo: string }>;
  salarios_suspeitos: Array<{ tipo: string; descricao: string; periodo: string }>;
  periodos_sem_remuneracao: unknown[];
}

export interface CalculoRMI {
  id: string;
  tenant_id: string;
  cnis_id: string;
  cliente_id: string;
  nome_calculo: string;
  tipo_beneficio: TipoBeneficio;
  regra_aplicada: string | null;
  data_der: string;
  idade_na_der: number;
  tempo_contribuicao_na_der: number;
  salario_beneficio: number;
  coeficiente_calculo: number;
  fator_previdenciario: number | null;
  rmi_calculada: number;
  rmi_teto: number | null;
  rmi_final: number;
  detalhamento_calculo: {
    passo_a_passo: string[];
    total_salarios_analisados: number;
    salario_beneficio: number;
    coeficiente_aplicado: number;
    fator_previdenciario: number | null;
    [key: string]: unknown;
  };
  requisitos_atendidos: Record<string, unknown> | null;
  rmi_regra_anterior: number | null;
  diferenca_reforma: number | null;
  calculo_valido: boolean;
  alertas: string[] | null;
  erros: string[] | null;
  versao_calculo: string | null;
  created_at: string;
}

export interface CalculoRMIRequest {
  cnis_id: string;
  tipo_beneficio: TipoBeneficio;
  data_der: string;
  nome_calculo: string;
  regra_aplicada?: string;
}

export interface Simulacao {
  id: string;
  cnis_id: string;
  nome_simulacao: string | null;
  data_simulacao_futura: string;
  taxa_crescimento_salario: number | null;
  taxa_inflacao_anual: number | null;
  idade_na_data: number | null;
  tempo_contribuicao_projetado: number | null;
  salario_beneficio_projetado: number | null;
  rmi_projetada: number | null;
  rmi_valor_atual: number | null;
  created_at: string;
}

export interface SimulacaoRequest {
  cnis_id: string;
  nome_simulacao?: string;
  data_simulacao_futura: string;
  taxa_crescimento_salario: number;
  taxa_inflacao_anual: number;
}

export type TipoRemuneracao =
  | "salario"
  | "13o_salario"
  | "ferias"
  | "rescisao"
  | "adicional_noturno"
  | "hora_extra"
  | "comissao"
  | "gratificacao"
  | "outros";

export const TIPOS_REMUNERACAO: Record<TipoRemuneracao, string> = {
  salario: "Salário",
  "13o_salario": "13º Salário",
  ferias: "Férias",
  rescisao: "Rescisão",
  adicional_noturno: "Adicional Noturno",
  hora_extra: "Hora Extra",
  comissao: "Comissão",
  gratificacao: "Gratificação",
  outros: "Outros",
};

export interface RemuneracaoItem {
  id: string;
  cnis_id: string;
  mes_referencia: string;
  ano: number;
  mes: number;
  salario_contribuicao: number;
  salario_contribuicao_corrigido: number | null;
  teto_inss: number | null;
  tipo_remuneracao: TipoRemuneracao | null;
  contribuiu_inss: boolean;
  salario_valido: boolean;
  acima_teto: boolean;
  abaixo_minimo: boolean;
}

export interface RemuneracoesListResponse {
  items: RemuneracaoItem[];
  total: number;
  skip: number;
  limit: number;
}

export interface RemuneracaoCreateRequest {
  competencia: string;
  salario_contribuicao: number;
  tipo_remuneracao: TipoRemuneracao;
  contribuiu_inss: boolean;
}

export interface ImportacaoCSVResult {
  criadas: number;
  ignoradas: number;
  erros: string[];
}

File: backend/app/schemas/cnis.py
import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, Field


# ─── Tipos compartilhados ────────────────────────────────────────────────────

TipoBeneficio = Literal[
    "aposentadoria_idade_urbana",
    "aposentadoria_idade_rural",
    "aposentadoria_tempo_contribuicao",
    "aposentadoria_especial_15",
    "aposentadoria_especial_20",
    "aposentadoria_especial_25",
    "aposentadoria_pcd_idade",
    "aposentadoria_pcd_tempo",
    "auxilio_doenca",
    "aposentadoria_invalidez",
    "salario_maternidade",
    "pensao_morte",
]

TipoContribuinte = Literal[
    "empregado", "empregador", "contribuinte_individual",
    "segurado_especial", "facultativo", "estagiario",
    "avulso", "dirigente_sindical",
]


# ─── CNIS ────────────────────────────────────────────────────────────────────

class CNISCreate(BaseModel):
    cliente_id: uuid.UUID
    nome_segurado: str = Field(..., min_length=2, max_length=255)
    cpf: str = Field(..., min_length=11, max_length=11)
    nis: str = Field(..., min_length=11, max_length=11)
    data_nascimento: date
    arquivo_original_nome: str | None = None


class CNISResponse(BaseModel):
    id: uuid.UUID
    tenant_id: uuid.UUID
    cliente_id: uuid.UUID
    nome_segurado: str
    nis: str
    data_nascimento: date
    periodo_inicial_cn: date | None
    periodo_final_cn: date | None
    tempo_contribuicao_total_dias: int | None
    tempo_contribuicao_anos: Decimal | None
    total_contribuicoes: int | None
    maior_salario_contribuicao: Decimal | None
    media_salarios_contribuicao: Decimal | None
    status_processamento: str
    erros_validacao: dict | None
    created_at: datetime

    model_config = {"from_attributes": True}


# ─── Período de Contribuição ─────────────────────────────────────────────────

class PeriodoContribuicaoCreate(BaseModel):
    cnis_id: uuid.UUID
    cnpj_empregador: str | None = None
    razao_social_empregador: str | None = None
    data_inicio: date
    data_fim: date | None = None
    tipo_contribuinte: TipoContribuinte | None = None
    categoria: str | None = None
    dias_contribuidos: int = Field(..., ge=0)
    indicador_aposentadoria_especial: bool = False
    agente_nocivo: str | None = None
    grau_deficiencia: Literal["leve", "moderada", "grave"] | None = None
    observacoes: str | None = None


class PeriodoContribuicaoResponse(BaseModel):
    id: uuid.UUID
    cnpj_empregador: str | None
    razao_social_empregador: str | None
    data_inicio: date
    data_fim: date | None
    tipo_contribuinte: str | None
    dias_contribuidos: int
    indicador_aposentadoria_especial: bool
    agente_nocivo: str | None
    grau_deficiencia: str | None
    total_remuneracoes: Decimal | None
    quantidade_remuneracoes: int | None
    periodo_valido: bool
    motivo_invalidacao: str | None

    model_config = {"from_attributes": True}


# ─── Remuneração ─────────────────────────────────────────────────────────────

class RemuneracaoResponse(BaseModel):
    id: uuid.UUID
    mes_referencia: date
    ano: int
    mes: int
    salario_contribuicao: Decimal
    salario_contribuicao_corrigido: Decimal | None
    teto_inss: Decimal | None
    contribuiu_inss: bool
    salario_valido: bool
    acima_teto: bool
    abaixo_minimo: bool

    model_config = {"from_attributes": True}


# ─── Cálculo RMI ─────────────────────────────────────────────────────────────

class CalculoRMIRequest(BaseModel):
    cnis_id: uuid.UUID
    tipo_beneficio: TipoBeneficio
    data_der: date = Field(..., description="Data de Entrada do Requerimento (real ou simulada)")
    nome_calculo: str = Field(..., min_length=2, max_length=100)
    regra_aplicada: str | None = None


class CalculoRMIResponse(BaseModel):
    id: uuid.UUID
    tenant_id: uuid.UUID
    cnis_id: uuid.UUID
    cliente_id: uuid.UUID
    nome_calculo: str
    tipo_beneficio: str
    regra_aplicada: str | None
    data_der: date
    idade_na_der: int
    tempo_contribuicao_na_der: Decimal
    salario_beneficio: Decimal
    coeficiente_calculo: Decimal
    fator_previdenciario: Decimal | None
    fator_acumulador: Decimal | None
    rmi_calculada: Decimal
    rmi_teto: Decimal | None
    rmi_final: Decimal
    detalhamento_calculo: dict
    requisitos_atendidos: dict | None
    rmi_regra_anterior: Decimal | None
    diferenca_reforma: Decimal | None
    calculo_valido: bool
    alertas: list | None
    erros: list | None
    versao_calculo: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


# ─── Simulação de Cenário ─────────────────────────────────────────────────────

class SimulacaoRequest(BaseModel):
    cnis_id: uuid.UUID
    nome_simulacao: str | None = None
    data_simulacao_futura: date
    taxa_crescimento_salario: Decimal = Field(Decimal("0.03"), ge=0, le=1)
    taxa_inflacao_anual: Decimal = Field(Decimal("0.045"), ge=0, le=1)


class SimulacaoResponse(BaseModel):
    id: uuid.UUID
    cnis_id: uuid.UUID
    nome_simulacao: str | None
    data_simulacao_futura: date
    taxa_crescimento_salario: Decimal | None
    taxa_inflacao_anual: Decimal | None
    idade_na_data: int | None
    tempo_contribuicao_projetado: Decimal | None
    salario_beneficio_projetado: Decimal | None
    rmi_projetada: Decimal | None
    rmi_valor_atual: Decimal | None
    created_at: datetime

    model_config = {"from_attributes": True}


# ─── Análise de Inconsistências ───────────────────────────────────────────────

class InconsistenciaItem(BaseModel):
    tipo: str
    descricao: str
    periodo_afetado: str | None = None
    impacto_financeiro: Decimal | None = None
    recomendacao: str | None = None


class AnaliseInconsistenciasResponse(BaseModel):
    cnis_id: uuid.UUID
    total_inconsistencias: int
    inconsistencias: list[InconsistenciaItem]
    periodos_sobrepostos: list[dict]
    salarios_suspeitos: list[dict]
    periodos_sem_remuneracao: list[dict]
